Lib/unittest/async_case.py

```diff
@@ -116,6 +116,7 @@ def _setupAsyncioRunner(self):
         assert self._asyncioRunner is None, 'asyncio runner is already initialized'
         runner = asyncio.Runner(debug=True)
         self._asyncioRunner = runner
+        asyncio.set_event_loop(runner.get_loop())
```

Comment:
So I'm guessing this can be reduced to `runner.get_loop()` (i.e. not calling `set_event_loop()`) since `get_loop()` forces the lazy initialization to happen which (once #95900 is fixed) will call `set_event_loop()` for us. Is that what you had in mind, @kumaraditya303 ?

Reply:
I would also like to understand why `TestCase1` run doesn't fail. Is it because it's getting the default loop created by `BaseDefaultEventLoopPolicy.get_event_loop()` when you call it before its `set_event_loop()` is ever called? (This would warrant a comment.)

Reply:
>  Is that what you had in mind,

Yes we don't have to call `set_event_loop` manually then.

Reply:
> I would also like to understand why TestCase1 run doesn't fail. Is it because it's getting the default loop created by BaseDefaultEventLoopPolicy.get_event_loop() when you call it before its set_event_loop() is ever called? (This would warrant a comment.)

`TestCase1` gets the default loop which is created on demand as no `set_event_loop` was ever called.

After `TestCase1 ` is completed, `set_event_loop` is called with `None`, so when `TestCase2` is run 'get_event_loop' in `setUp` checks if `set_event_loop` was called with `None` and no other `set_event_loop` with `loop` was called so it raises error.

https://github.com/python/cpython/blob/914f6367a0d015986dafa7a9d542e24192753b6b/Lib/asyncio/events.py#L666-L680

Reply:
Is this test relying on the implicit global state? Of the default policy

Eg if you pasted this test twice would it behave differently?

Do you need to make sure `set_event_loop_policy(None)` is called before this test rather than just the module?

Reply:
IIUC the code added by this PR cals `runner.get_loop()`, which calls `runner._lazy_init()`, which calls `set_event_loop()`, so we don't rely on the default policy giving us a "free" new event loop on the first `get_event_loop()` call. This fix depends on the fact that `_setupAsyncioRunner()` doesn't pass `loop_factory=...` when it constructs the runner a few lines before, because otherwise `_lazy_init()` wouldn't call `set_event_loop()`.

Note that the offending `get_event_loop()` call is in the test itself:
```py
                 self.loop = asyncio.get_event_loop_policy().get_event_loop()
```
(This is the reproducer you came up with yourself in #95736.)

The problem that this PR is *fixing* was discovered only when two tests were run, since the first test ended up calling `get_event_loop()` before the runner initialized so it got a "free" new event loop from the default policy, but the second test was not so lucky.

So I don't think we need to reset the event loop policy before this test. But I may be missing a step -- if so please show what I am missing.

Reply:
It does seem that the inner test, `TestCase1`/`TestCase2` (like in your reproducer) needs to be repeated twice because the first time the `get_event_loop()` calls succeeds due to the default policy. Maybe we can avoid this duplication by inserting `set_event_loop(None)` at the top of the test (or at least before calling `test.debug()`). @kumaraditya303 Do you follow what I am suggesting?

Reply:
> set_event_loop(None)

This is what we do in distributed fwiw

Reply:
> So I don't think we need to reset the event loop policy before this test. But I may be missing a step -- if so please show what I am missing.

Ah I see what you mean. In the *fixed* code there's no longer any global state, but in the code in main there would have been.

My reproducer (built from deleting everything in aiohttp that didn't cause the problem) only works because of the bug, however `set_event_loop(None)` just does the right thing regardless of the global state of the policy (as long as it is the default policy) in spite of the bug?